lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
+    // Generate a popup window
+    // TODO: Refactor this so that openWindow throws an error, loginPopupHelper rejects or resolves based on that action
+    var popUpWindow = this.openWindow("about:blank", "_blank", 1, this, resolve, reject);
+    if (!popUpWindow) {
+      return;
+    }
 
-      this.authorityInstance.ResolveEndpointsAsync().then(() => {
-          const authenticationRequest = new AuthenticationRequestParameters(this.authorityInstance, this.clientId, scopes, ResponseTypes.id_token, this.getRedirectUri(), this._state);
-          if (extraQueryParameters) {
-              authenticationRequest.extraQueryParameters = extraQueryParameters;
-          }
+    // Track login progress
+    this._loginInProgress = true;
 
-          this._cacheStorage.setItem(Constants.loginRequest, window.location.href, this.storeAuthStateInCookie);
-          this._cacheStorage.setItem(Constants.loginError, "");
-          this._cacheStorage.setItem(Constants.nonceIdToken, authenticationRequest.nonce, this.storeAuthStateInCookie);
-          this._cacheStorage.setItem(Constants.msalError, "");
-          this._cacheStorage.setItem(Constants.msalErrorDescription, "");
-          const authorityKey = Constants.authority + Constants.resourceDelimeter + authenticationRequest.state;
-          this._cacheStorage.setItem(authorityKey, this.authority, this.storeAuthStateInCookie);
-          const urlNavigate = authenticationRequest.createNavigateUrl(scopes)  + Constants.response_mode_fragment;
-          window.renewStates.push(authenticationRequest.state);
-          window.requestType = Constants.login;
-          this.registerCallback(authenticationRequest.state, scope, resolve, reject);
-          if (popUpWindow) {
-              this._logger.infoPii("Navigated Popup window to:" + urlNavigate);
-              popUpWindow.location.href = urlNavigate;
-          }
+    // Resolve endpoint
+    this.authorityInstance.ResolveEndpointsAsync().then(() => {
+      const authenticationRequest = new AuthenticationRequestParameters(this.authorityInstance, this.clientId, scopes, ResponseTypes.id_token, this.getRedirectUri(), this._state);
+      
+      // set extraQueryParameters
+      if (extraQueryParameters) {
+        authenticationRequest.extraQueryParameters = extraQueryParameters;
+      }
 
-      }, () => {
-          this._logger.info(ErrorCodes.endpointResolutionError + ":" + ErrorDescription.endpointResolutionError);
-          this._cacheStorage.setItem(Constants.msalError, ErrorCodes.endpointResolutionError);
-          this._cacheStorage.setItem(Constants.msalErrorDescription, ErrorDescription.endpointResolutionError);
-          if (reject) {
-              reject(ErrorCodes.endpointResolutionError + ":" + ErrorDescription.endpointResolutionError);
-          }
+      // Cache the state, nonce, and login request data
+      this._cacheStorage.setItem(Constants.loginRequest, window.location.href, this.storeAuthStateInCookie);
```

Comment:
same question about setting constants, also this looks to be repeated code, can it be factored

Reply:
See above about setting constants. this code should be extracted to a helper function, I believe this will be in a future PR.